### PR TITLE
MM-16819 Helper function to build path to plugin asset

### PR DESCRIPTION
**Pull request by Lumexralph**
#### Summary

* This PR contains the implementation of a function that helps a plugin build a URL to a resource or custom assets via the assets directory

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/13588

### CODE REVIEW

File: plugin/helpers_plugin.go [hunk 1]
```diff
@@ -54,3 +54,16 @@ func (p *HelpersImpl) ensureServerVersion(required string) error {
 	}
 	return nil
 }
+
+// GetPluginAssetURL builds a URL to the resource from asset directory.
+func (p *HelpersImpl) GetPluginAssetURL(pluginID, asset string) (*url.URL, error) {
```
Comment (lieut-data) [inline plugin/helpers_plugin.go]: Can we add this function to the `Helpers` interface? Without the change to the interface, no plugin can actually use this method .

File: plugin/helpers_plugin.go [hunk 2]
```diff
@@ -54,3 +54,16 @@ func (p *HelpersImpl) ensureServerVersion(required string) error {
...
+	if len(pluginID) == 0 || len(asset) == 0 {
+		return nil, errors.New("empty pluginID or directory name provided")
```
Comment (lieut-data) [inline plugin/helpers_plugin.go]: Let's break these checks apart into discrete errors to make it clear to the plugin author what went wrong. Also, I think it would be an empty `asset name` not `directory name`?

File: plugin/helpers_plugin.go [hunk 3]
```diff
@@ -54,3 +54,16 @@ func (p *HelpersImpl) ensureServerVersion(required string) error {
...
+	}
+	siteURL := *p.API.GetConfig().ServiceSettings.SiteURL
```
Comment (lieut-data) [inline plugin/helpers_plugin.go]: style nit: let's add a newline before this line, to group things a little more clearly.

Comment (lieut-data) [reply plugin/helpers_plugin.go]: There's a tiny chance that `SiteURL` is an empty string -- suggest we check for that and also return an error.

File: plugin/helpers_plugin_test.go [hunk 4]
```diff
@@ -103,3 +104,30 @@ func TestInstallPluginFromURL(t *testing.T) {
 		assert.Equal(t, "received 404 status code while downloading plugin from server", err.Error())
 	})
 }
+
+func TestGetPluginAssetURL(t *testing.T) {
```
Comment (lieut-data) [inline plugin/helpers_plugin_test.go]: This is a great unit test, but as noted above, won't catch the case of a plugin trying to compile to actually use this method and finding it missing from the interface. Can you replicate something like this over in `app/plugin_api_test.go`, where you'll note we actually compile plugins on demand to exercise? You might need to setup a slightly different pattern to test the helpers, but all the foundational bits are there.

File: plugin/helpers_plugin_test.go [hunk 5]
```diff
@@ -103,3 +104,30 @@ func TestInstallPluginFromURL(t *testing.T) {
...
+		pluginID := "mattermost-1234"
+		dir := "assets"
+		wantedURL := &url.URL{Scheme: "https", Host: "mattermost.example.com", Path: "/mattermost-1234/assets"}
+		gotURL, err := p.GetPluginAssetURL(pluginID, dir)
+
+		assert.Equalf(t, wantedURL, gotURL, "GetPluginAssetURL(%s, %s) got=%s; want=%v", pluginID, dir, gotURL, wantedURL)
+		assert.NoError(t, err)
+	})
+
+	t.Run("Invalid directory and pluginID provided", func(t *testing.T) {
```
Comment (lieut-data) [inline plugin/helpers_plugin_test.go]: Should we add separate tests cases for invalid directory and invalid pluginID? (And then, based on my comment above, for when SiteURL is empty?)

File: plugin/helpers_plugin.go (hunk 1 above)
Comment (Lumexralph) [reply plugin/helpers_plugin.go]: @lieut-data I will work on that, thanks for the feedback.

File: plugin/helpers_plugin.go (hunk 2 above)
Comment (Lumexralph) [reply plugin/helpers_plugin.go]: @lieut-data Cool! Thanks for the feedback, I will work on it. I think it'll be better to split the errors for different scenarios.

File: plugin/helpers_plugin.go (hunk 3 above)
Comment (Lumexralph) [reply plugin/helpers_plugin.go]: Thanks for the thoughtful feedback, I'll work on them.

File: plugin/helpers_plugin_test.go (hunk 4 above)
Comment (Lumexralph) [reply plugin/helpers_plugin_test.go]: @lieut-data Okay, I will work on that and reach out if I need any help on it.

File: plugin/helpers_plugin_test.go (hunk 5 above)
Comment (Lumexralph) [reply plugin/helpers_plugin_test.go]: @lieut-data What criteria make a directory and pluginID invalid? can you please talk a bit on this?

Comment (lieut-data) [reply plugin/helpers_plugin_test.go]: I was thinking specifically the case of `len(pluginID) == 0` and `len(asset) == 0`, vs. testing them both at the same time, when we're really only hitting one branch of the if condition.

File: plugin/helpers_plugin_test.go (hunk 4 above)
Comment (Lumexralph) [reply plugin/helpers_plugin_test.go]: @lieut-data I have implemented all the feedback but this part is not clear yet to me, would you be kind to help out?

File: plugin/helpers_plugin_test.go (hunk 5 above)
Comment (Lumexralph) [reply plugin/helpers_plugin_test.go]: I have implemented all the feedback.

File: plugin/helpers.go [hunk 6]
```diff
@@ -78,6 +80,11 @@ type Helpers interface {
 	//
 	// Minimum server version: 5.18
 	InstallPluginFromURL(downloadURL string, replace bool) (*model.Manifest, error)
+
+	// GetPluginAssetURL builds a URL to the resource from asset directory.
+	// Plugins may expose custom assets via the assets directory,
+	// but must then combine the SiteURL and their plugin ID to build the resulting path.
```
Comment (lieut-data) [inline plugin/helpers.go]: This text is helpful in the context of the PR, but I think is confusing here, since it sounds like a requirement on the part of the plugin author. Instead, this helper method simplifies doing that work. Propose:
```suggestion
	// GetPluginAssetURL builds a URL to the given asset in the assets directory.
	// Use this URL to link to assets from the webapp, or for third-party integrations with your plugin.
```

File: plugin/helpers_plugin.go [hunk 7]
```diff
@@ -57,10 +57,18 @@ func (p *HelpersImpl) ensureServerVersion(required string) error {
 
 // GetPluginAssetURL builds a URL to the resource from asset directory.
```
Comment (lieut-data) [inline plugin/helpers_plugin.go]:
```suggestion
// GetPluginAssetURL builds a URL to the given asset in the assets directory.
```

File: plugin/helpers_plugin.go [hunk 8]
```diff
@@ -57,10 +57,18 @@ func (p *HelpersImpl) ensureServerVersion(required string) error {
...
+		return nil, errors.New("empty pluginID provided")
 	}
+
+	if len(asset) == 0 {
+		return nil, errors.New("empty asset name provided")
+	}
+
 	siteURL := *p.API.GetConfig().ServiceSettings.SiteURL
+	if siteURL == "" {
+		return nil, errors.New("No SiteURL found in the service settings")
```
Comment (lieut-data) [inline plugin/helpers_plugin.go]: Errors in Go usually start with a lowercase letter, since they are often composed into a larger construct: `something failed: in this code: no SiteURL configured by the server`. I'd also suggest tweaking the error as such:
```suggestion
		return nil, errors.New("no SiteURL configured by the server")
```

File: plugin/helpers_plugin_test.go [hunk 9]
```diff
@@ -112,22 +112,53 @@ func TestGetPluginAssetURL(t *testing.T) {
...
+		gotURL, err := p.GetPluginAssetURL(pluginID, dir)
+
+		assert.Nilf(t, gotURL, "GetPluginAssetURL(%q, %q) got=%q; want=nil", pluginID, dir, gotURL)
+		assert.Error(t, err)
+	})
+
+	siteURL = ""
+	api.On("GetConfig").Return(&model.Config{ServiceSettings: model.ServiceSettings{SiteURL: &siteURL}})
+
+	t.Run("Empty SiteURL was provided", func(t *testing.T) {
```
Comment (lieut-data) [inline plugin/helpers_plugin_test.go]:
```suggestion
	t.Run("Empty SiteURL was configured", func(t *testing.T) {
```

File: plugin/helpers.go (hunk 6 above)
Comment (Lumexralph) [reply plugin/helpers.go]: @lieut-data Thanks for the suggestion, I will work on it.

File: plugin/helpers_plugin_test.go (hunk 9 above)
Comment (Lumexralph) [reply plugin/helpers_plugin_test.go]: @lieut-data  I have implemented the feedback

File: plugin/helpers_plugin.go [hunk 10]
```diff
@@ -54,3 +54,24 @@ func (p *HelpersImpl) ensureServerVersion(required string) error {
...
+	}
+
+	if len(asset) == 0 {
+		return nil, errors.New("empty asset name provided")
+	}
+
+	siteURL := *p.API.GetConfig().ServiceSettings.SiteURL
+	if siteURL == "" {
+		return nil, errors.New("no SiteURL configured by the server")
+	}
```
Comment (hanzei) [inline plugin/helpers_plugin.go]: Nit: Add a newline to separate the blocks.
```suggestion
	}
	
```

File: plugin/helpers.go [hunk 11]
```diff
@@ -78,6 +80,10 @@ type Helpers interface {
 	//
 	// Minimum server version: 5.18
 	InstallPluginFromURL(downloadURL string, replace bool) (*model.Manifest, error)
+
+	// GetPluginAssetURL builds a URL to the given asset in the assets directory.
+	// Use this URL to link to assets from the webapp, or for third-party integrations with your plugin.
+	GetPluginAssetURL(pluginID, asset string) (*url.URL, error)
```
Comment (hanzei) [inline plugin/helpers.go]: @Lumexralph and I have both looked into how passing the `pluginID` can be avoided and both didn't fond a way `HelpersImpl` could get `pluginID`. Do you see a way?

Comment (lieut-data) [reply plugin/helpers.go]: At present, I don't know of a good way to do this myself.

File: plugin/helpers_plugin.go (hunk 10 above)
Comment (Lumexralph) [reply plugin/helpers_plugin.go]: @hanzei thanks for the helpful feedback, I have implemented them.

File: plugin/helpers_plugin.go [hunk 12]
```diff
@@ -54,3 +54,26 @@ func (p *HelpersImpl) ensureServerVersion(required string) error {
 	}
 	return nil
 }
+
+// GetPluginAssetURL builds a URL to the given asset in the assets directory.
```
Comment (hanzei) [inline plugin/helpers_plugin.go]:
```suggestion
// GetPluginAssetURL implements GetPluginAssetURL.
```

File: plugin/helpers.go [hunk 13]
```diff
@@ -78,6 +80,10 @@ type Helpers interface {
 	//
 	// Minimum server version: 5.18
 	InstallPluginFromURL(downloadURL string, replace bool) (*model.Manifest, error)
+
+	// GetPluginAssetURL builds a URL to the given asset in the assets directory.
+	// Use this URL to link to assets from the webapp, or for third-party integrations with your plugin.
```
Comment (hanzei) [inline plugin/helpers.go]:
```suggestion
	// Use this URL to link to assets from the webapp, or for third-party integrations with your plugin.
	//
	// Minimum server version: 5.2
```

File: plugin/helpers_plugin.go (hunk 12 above)
Comment (lieut-data) [reply plugin/helpers_plugin.go]: If we go with this comment, I'd suggest omitting it altogether, since it's of no use to the developer. I'm personally ok with "duplicating" comments like this.

Comment (hanzei) [reply plugin/helpers_plugin.go]: There has been issues in the past where the comment on the interface and the implementation diverged which is a situation I want to avoid in the future.

I'm 2/5 that having this comment is better then having non because it makes clear that the comment was not forgotten but instead lives somewhere else.

Comment (Lumexralph) [reply plugin/helpers_plugin.go]: @lieut-data @hanzei What's the final resolution on this?

Comment (lieut-data) [reply plugin/helpers_plugin.go]: Let's adopt @hanzei's proposal here and be consistent with the rest of the helpers -- thanks for the discussion :)

**Comment by Lumexralph** [comment]
> I missed two carves in my review, sorry

@hanzei I have implemented the feedback, thanks.

**Comment by hanzei** [comment]
@prapti Heads up that this is PR is best tested by just using it in a plugin. Let me know, if I can help with anything.

**Comment by lieut-data** [comment]
@prapti fully agree. Can we use [`path.Join`](https://golang.org/pkg/path/#Join) to get the desired behaviour easily?

**Comment by Lumexralph** [comment]
> This is great @Lumexralph!
> One condition I ran into though, was when the directory path provided starts with a `/` (for example "/Users/prapti/path/to/matterpoll/assets") it generates a URL that looks like the following with double `//` - "http://localhost:8065/matterpoll//Users/prapti/path/to/matterpoll/assets". I feel like this case should be addressed when someone provides a valid path to the asset directory to parse the beginning `/` correctly that may otherwise cause issues with the generated URL.
> @hanzei, @lieut-data please feel free to add your thoughts.

@prapti Cool! I didn't think of that happening, I will work on the feedback.
@lieut-data I will try that out.

**Comment by Lumexralph** [comment]
@prapti I have worked on the feedback and also added a test for it. Please kindly review.

File: plugin/helpers.go [hunk 14]
```diff
@@ -85,7 +83,7 @@ type Helpers interface {
 	// Use this URL to link to assets from the webapp, or for third-party integrations with your plugin.
 	//
 	// Minimum server version: 5.2
-	GetPluginAssetURL(pluginID, asset string) (*url.URL, error)
+	GetPluginAssetURL(pluginID, asset string) (string, error)
```
Comment (lieut-data) [inline plugin/helpers.go]: 0/5, but why the change to `string` here from `url.URL`?

Comment (Lumexralph) [reply plugin/helpers.go]: @lieut-data `path.Join()` would require a `string`, I also figured out in the course of working on the feedback that it would easier and also for testing to return a string since what we care about is the `URL`, I didn't figure that initially when I worked on it. Is there any concern with the new implementation?

Comment (lieut-data) [reply plugin/helpers.go]: Sounds reasonable, thanks for clarifying @Lumexralph!

**Comment by jasonblais** [comment]
@prapti Friendly reminder to help with QA review, thanks!

**Comment by hanzei** [comment]
@Lumexralph Thanks for the contribution and sticking with the community :+1:

**Comment by Lumexralph** [comment]
> @Lumexralph Thanks for the contribution and sticking with the community 👍

Thanks for the opportunity to work with mattermost and always glad to help.
